File: functions/src/scoring/urgencyCalculator.js
/**
 * Urgency Score Calculator
 *
 * Computes a 0-100 urgency score for blood requests.
 * This determines priority queue position.
 *
 * Score Breakdown:
 * - Patient condition:  0-40 pts (critical=40, urgent=25, standard=10)
 * - Surgery window:     0-30 pts (closer = higher)
 * - Blood type rarity:  0-20 pts (AB-=20, B-/A-=15, O-=10, etc.)
 * - Units needed:       0-10 pts (2 pts per unit, max 10)
 */

function computeUrgency(req) {
  let score = 0;

  // Condition weight (0-40 pts)
  if (req.condition === 'critical') score += 40;
  else if (req.condition === 'urgent') score += 25;
  else score += 10;

  // Surgery window (0-30 pts)
  if (req.surgeryTime) {
    const surgeryMs = req.surgeryTime.toMillis
      ? req.surgeryTime.toMillis()
      : new Date(req.surgeryTime).getTime();
    const hoursLeft = (surgeryMs - Date.now()) / 3600000;
    score += Math.max(0, Math.min(30, 30 - hoursLeft * 3));
  }

  // Blood type rarity (0-20 pts)
  const rarityMap = {
    'AB-': 20,
    'B-': 15,
    'A-': 15,
    'O-': 10,
    'AB+': 8,
    'B+': 5,
    'A+': 5,
    'O+': 3,
  };
  score += rarityMap[req.bloodGroup] || 5;

  // Units needed (0-10 pts)
  score += Math.min(10, (req.units || 1) * 2);

  return Math.min(100, Math.round(score));
}

module.exports = { computeUrgency };

File: lib/utils/urgency_calculator.dart
import '../config/constants.dart';

/// Client-side urgency score calculator.
/// Mirrors the Cloud Function `computeUrgency` logic for preview.
class UrgencyCalculator {
  /// Compute urgency score (0-100) based on request parameters.
  static int compute({
    required String condition,
    required String bloodGroup,
    required int units,
    DateTime? surgeryTime,
  }) {
    int score = 0;

    // Condition weight (0-40 pts)
    switch (condition) {
      case 'critical':
        score += 40;
        break;
      case 'urgent':
        score += 25;
        break;
      default:
        score += 10;
    }

    // Surgery window (0-30 pts)
    if (surgeryTime != null) {
      final hoursLeft = surgeryTime.difference(DateTime.now()).inMinutes / 60.0;
      score += (30 - hoursLeft * 3).clamp(0, 30).round();
    }

    // Blood type rarity (0-20 pts)
    score += AppConstants.bloodGroupRarity[bloodGroup] ?? 5;

    // Units needed (0-10 pts)
    score += (units * 2).clamp(0, 10);

    return score.clamp(0, 100);
  }

  /// Human-readable urgency label.
  static String label(int score) {
    if (score >= 70) return 'Critical';
    if (score >= 40) return 'Urgent';
    return 'Standard';
  }

  /// Whether this score qualifies as emergency priority.
  static bool isEmergency(int score) => score >= 70;
}
